### PR TITLE
[Bug 17552] Fix crash in byteOffset() when string_to_search is empty

### PR DESCRIPTION
**Pull request by livecodepanos**

### CODE REVIEW

**Comment by peter-b** [comment]
Please add a regression test!

File: engine/src/exec-strings.cpp [hunk 1]
```diff
@@ -1919,7 +1919,7 @@ void MCStringsEvalByteOffset(MCExecContext& ctxt, MCDataRef p_chunk, MCDataRef p
     // SN-2014-09-05: [[ Bug 13346 ]] byteOffset is 0 if the byte is not found, and 'empty'
     // is by definition not found; getting in the loop ensures at least 1 is returned.
     
-    if (!MCDataIsEmpty(p_chunk) && MCDataFirstIndexOf(p_string, p_chunk, MCRangeMake(p_start_offset, UINDEX_MAX), t_result))
+    if (!MCDataIsEmpty(p_chunk) && !MCDataIsEmpty(p_string) && MCDataFirstIndexOf(p_string, p_chunk, MCRangeMake(p_start_offset, UINDEX_MAX), t_result))
```
Comment (runrevmark) [inline engine/src/exec-strings.cpp]: I think the bug here is actually in MCDataFirstIndexOf - I think it should return false if MCDataGetLength(p_chunk) == 0 or MCDataGetLength(p_string) == 0 - as I believe that is what MCStringFirstIndexOf does.

Comment (peter-b) [inline engine/src/exec-strings.cpp]: @runrevmark I believe that's correct; no byte can be found in the empty bytestring. We should add a test to the LCB test suite to cover that.

File: libfoundation/src/foundation-data.cpp [hunk 2]
```diff
@@ -779,11 +779,16 @@ bool MCDataFirstIndexOf(MCDataRef p_data, MCDataRef p_chunk, MCRange t_range, ui
 {
 	__MCAssertIsData(p_data);
 	__MCAssertIsData(p_chunk);
+	
+	if (MCDataGetLength(p_data) == 0)
+		return false;
```
Comment (runrevmark) [inline libfoundation/src/foundation-data.cpp:786]: I suggest removing this check and replacing it with a t_range.length == 0 check after the __MCDataClampRange call.

File: tests/lcs/core/strings/byteoffset.livecodescript [hunk 3]
```diff
@@ -0,0 +1,36 @@
+script "CoreStringsByteoffset"
+/*
```
Comment (runrevmark) [inline tests/lcs/core/strings/byteoffset.livecodescript:2]: I suggest adding explicit tests for empty needle in empty haystack, and empty needle in non-empty haystack too.

**Comment by runrevmark** [comment]
@livecode-vulcan review ok c49191a

**Comment by livecode-vulcan** [comment]
💙 review by @runrevmark ok c49191a29ad6251098994c9e4672e44afa5728d9

**Comment by livecode-vulcan** [comment]
😞 test failure c49191a29ad6251098994c9e4672e44afa5728d9
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/2476)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/2486)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/2469)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/2487)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/2493)
- try-community-win-x86: [failure](https://vulcan.livecode.com/builders/try-community-win-x86/builds/2497)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/2509)

**Comment by livecodepanos** [comment]
Failure because of Windows timeout, so:

@livecode-vulcan retry c49191a

**Comment by livecode-vulcan** [comment]
😎 test success c49191a29ad6251098994c9e4672e44afa5728d9
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/2478)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/2488)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/2471)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/2489)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/2495)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/2499)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/2511)
